src/utils/offlineRequestHandler.ts

```diff
@@ -0,0 +1,348 @@
+    request: Request,
+    type: string,
+    stepKey?: string
+  ): Promise<void> {
+    const db = await this.openDatabase();
+    try {
+      const getTransaction = async () => {
+        const tx = db.transaction(this.storeName, 'readwrite');
+        const store = tx.objectStore(this.storeName);
+        return { tx, store };
+      };
+      const requestClone: Request = request.clone();
+      let serializedBody: SerializedRequestBody = { type: 'text', body: '' };
+
+      if (requestClone.method !== 'GET' && requestClone.method !== 'HEAD') {
+        serializedBody = await requestClone
+          .blob()
+          .then(this.serializeRequestBody);
+      }
+
+      const headers: Record<string, string> = {};
+      request.headers.forEach((value, key) => {
+        headers[key] = value;
+      });
+
+      const serializedRequest: SerializedRequest = {
+        url: requestClone.url,
+        method: requestClone.method,
+        headers: JSON.stringify(headers),
+        body: serializedBody.body,
+        bodyType: serializedBody.type,
+        timestamp: Date.now(),
+        type,
+        stepKey
+      };
+      const { tx, store } = await getTransaction();
+      const requestPromise = new Promise<void>((resolve, reject) => {
+        tx.oncomplete = () => resolve();
+        tx.onerror = () => reject(tx.error);
+      });
+
+      store.add(serializedRequest);
+      await requestPromise;
+      console.log('Request successfully added: ', request, type, stepKey);
+    } catch (error) {
+      console.error('Error adding request to IndexedDB:', error);
+    }
+  }
+
+  // Helper function to serialize request body based on its type
+  private async serializeRequestBody(
+    body: any
+  ): Promise<SerializedRequestBody> {
+    if (body === null) {
+      return { type: 'text', body: '' };
+    }
+    if (body instanceof Blob) {
+      return { type: 'blob', body: await body.arrayBuffer() }; // Convert Blob to ArrayBuffer for storage
+    } else if (body instanceof FormData) {
+      // FormData needs to be converted to a plain object
+      const formDataObj: Record<string, any> = {};
+      body.forEach((value, key) => {
+        formDataObj[key] = value;
+      });
+      return { type: 'formData', body: formDataObj };
+    } else if (body instanceof ArrayBuffer) {
+      return { type: 'arrayBuffer', body: body };
+    } else {
+      // Assume it's text or JSON
+      return { type: 'text', body: await body.text() };
+    }
+  }
+
+  // TODO: what does performance look like while using IndexDB?
+  // Replay requests from IndexedDB
+  public async replayRequests() {
+    if (!navigator.onLine) {
+      // If offline, block until requests are replayed from online event handler
+      await new Promise((resolve) => this._addOnlineSignal(resolve));
+      return;
+    }
+
+    if (this.isReplayingRequests) {
+      return;
+    }
+
+    this.isReplayingRequests = true;
+
+    try {
+      const db = await this.openDatabase();
```

Comment:
please unify duplicated code. in this case, the db transaction initialization code can probably all be unified into a method.